text/0000-bitfields.md

```diff
@@ -0,0 +1,340 @@
+    #[bitfield(1)] pub exclude_user: u64,
+    #[bitfield(1)] pub exclude_kernel: u64,
+    #[bitfield(1)] pub exclude_hv: u64,
+    #[bitfield(1)] pub exclude_idle: u64,
+    #[bitfield(1)] pub mmap: u64,
+    #[bitfield(1)] pub comm: u64,
+    #[bitfield(1)] pub freq: u64,
+    #[bitfield(1)] pub inherit_stat: u64,
+    #[bitfield(1)] pub enable_on_exec: u64,
+    #[bitfield(1)] pub task: u64,
+    #[bitfield(1)] pub watermark: u64,
+    #[bitfield(1)] pub precise_ip: u64,
+    #[bitfield(1)] pub mmap_data: u64,
+    #[bitfield(1)] pub sample_id_all: u64,
+    #[bitfield(1)] pub exclude_host: u64,
+    #[bitfield(1)] pub exclude_guest: u64,
+    #[bitfield(1)] pub exclude_callchain_kernel: u64,
+    #[bitfield(1)] pub exclude_callchain_user: u64,
+    #[bitfield(1)] pub mmap2: u64,
+    #[bitfield(1)] pub comm_exec: u64,
+    #[bitfield(39)] pub __reserved_1: u64,
+
+    /* ... */
+}
+```
+
+`N` must evaluate to an integer in the range `[0, sizeof(T)]` where `T` is the
+declared type of the field. It specifies the width of the field.
+
+Many details of the behavior of such fields are left to the implementation.
+However, the implementation is encouraged to follow the behavior of the dominant
+C ABI on the target platform.
+
+Those details that are specified are taken from the C11 standard and reproduced
+here for completeness. However, some of the details specified in said standard
+are omitted here since they are implied by the `#[repr(C)]` attribute and thus
+already covered by the previous paragraph.
+
+Concurrent access to different bitfields inside of an object is a data race
+unless,
+
+* none of the accesses mutate the bitfields,
+* the accesses are synchronized,
+* the bitfields are separated (in declaration order) by a non-bitfield, or
+* the bitfields are separated (in declaration order) by a bitfield of size 0.
+
+For example:
+
+```rust
+#[repr(C)]
+struct X {
+    #[bitfield(8)] pub a: u8,
+    #[bitfield(8)] pub b: u8,
+    #[bitfield(0)] pub c: u8,
+    #[bitfield(1)] pub d: u8,
+}
+```
+
+In this structure, `a` and `b` must not be modified concurrently. `a` and `d`
+can be accessed concurrently.
+
+Values stored in a bitfield have a binary representation using the `m` bits it
+occupies in its memory location.
+
+The `&` and `&mut` operators cannot be applied to bitfields.
+
+The type of a bitfield is its declared type. Reading from a bitfield returns a
+value of the declared type. Only values of the declared type can be written to
+a bitfield. The bits of the binary representation of a bitfield are interpreted
+as the bits of a signed or unsigned integer of the width of the bitfield. In
+particular, a signed bitfield of non-zero size contains a sign bit. When a
+zero-sized bitfield is read, `0` is returned.
+
+For the purpose of overflow checking, bitfields are treated as signed or
+unsigned integers of their declared width. In this context, the only value that
+should be written to a bitfield of size `0` is `0`.
```

Comment:
What happens in non-debug builds where we don't panic on overflow? Is the write just ignored?